<!DOCTYPE HTML>
<html lang="en" class="sidebar-visible no-js light">
    <head>
        <!-- Book generated using mdBook -->
        <meta charset="UTF-8">
        <title>10X Alignment - How Do I...?</title>
        <!-- Custom HTML head -->
        <meta content="text/html; charset=utf-8" http-equiv="Content-Type">
        <meta name="description" content="">
        <meta name="viewport" content="width=device-width, initial-scale=1">
        <meta name="theme-color" content="#ffffff" />

        <link rel="icon" href="../favicon.svg">
        <link rel="shortcut icon" href="../favicon.png">
        <link rel="stylesheet" href="../css/variables.css">
        <link rel="stylesheet" href="../css/general.css">
        <link rel="stylesheet" href="../css/chrome.css">
        <link rel="stylesheet" href="../css/print.css" media="print">
        <!-- Fonts -->
        <link rel="stylesheet" href="../FontAwesome/css/font-awesome.css">
        <link rel="stylesheet" href="../fonts/fonts.css">
        <!-- Highlight.js Stylesheets -->
        <link rel="stylesheet" href="../highlight.css">
        <link rel="stylesheet" href="../tomorrow-night.css">
        <link rel="stylesheet" href="../ayu-highlight.css">

        <!-- Custom theme stylesheets -->
    </head>
    <body>
        <!-- Provide site root to javascript -->
        <script type="text/javascript">
            var path_to_root = "../";
            var default_theme = window.matchMedia("(prefers-color-scheme: dark)").matches ? "navy" : "light";
        </script>

        <!-- Work around some values being stored in localStorage wrapped in quotes -->
        <script type="text/javascript">
            try {
                var theme = localStorage.getItem('mdbook-theme');
                var sidebar = localStorage.getItem('mdbook-sidebar');

                if (theme.startsWith('"') && theme.endsWith('"')) {
                    localStorage.setItem('mdbook-theme', theme.slice(1, theme.length - 1));
                }

                if (sidebar.startsWith('"') && sidebar.endsWith('"')) {
                    localStorage.setItem('mdbook-sidebar', sidebar.slice(1, sidebar.length - 1));
                }
            } catch (e) { }
        </script>

        <!-- Set the theme before any content is loaded, prevents flash -->
        <script type="text/javascript">
            var theme;
            try { theme = localStorage.getItem('mdbook-theme'); } catch(e) { }
            if (theme === null || theme === undefined) { theme = default_theme; }
            var html = document.querySelector('html');
            html.classList.remove('no-js')
            html.classList.remove('light')
            html.classList.add(theme);
            html.classList.add('js');
        </script>

        <!-- Hide / unhide sidebar before it is displayed -->
        <script type="text/javascript">
            var html = document.querySelector('html');
            var sidebar = 'hidden';
            if (document.body.clientWidth >= 1080) {
                try { sidebar = localStorage.getItem('mdbook-sidebar'); } catch(e) { }
                sidebar = sidebar || 'visible';
            }
            html.classList.remove('sidebar-visible');
            html.classList.add("sidebar-" + sidebar);
        </script>

        <nav id="sidebar" class="sidebar" aria-label="Table of contents">
            <div class="sidebar-scrollbox">
                <ol class="chapter"><li class="chapter-item expanded affix "><a href="../intro.html">Introduction</a></li><li class="spacer"></li><li class="chapter-item expanded "><a href="../crispr_screen/intro.html"><strong aria-hidden="true">1.</strong> CRISPR Screening</a></li><li><ol class="section"><li class="chapter-item expanded "><a href="../crispr_screen/abundance.html"><strong aria-hidden="true">1.1.</strong> Counting sgRNA abundance</a></li><li><ol class="section"><li class="chapter-item expanded "><a href="../crispr_screen/reference.html"><strong aria-hidden="true">1.1.1.</strong> Reference Library</a></li><li class="chapter-item expanded "><a href="../crispr_screen/counting.html"><strong aria-hidden="true">1.1.2.</strong> Guide Counting</a></li></ol></li><li class="chapter-item expanded "><a href="../crispr_screen/expression.html"><strong aria-hidden="true">1.2.</strong> Differential Expression</a></li><li class="chapter-item expanded "><a href="../crispr_screen/visualization.html"><strong aria-hidden="true">1.3.</strong> Visualization</a></li></ol></li><li class="chapter-item expanded "><a href="../cropseq/intro.html"><strong aria-hidden="true">2.</strong> CROP-Seq</a></li><li><ol class="section"><li class="chapter-item expanded "><a href="../cropseq/installation.html"><strong aria-hidden="true">2.1.</strong> Installation</a></li><li class="chapter-item expanded "><a href="../cropseq/environment.html"><strong aria-hidden="true">2.2.</strong> Environment</a></li><li class="chapter-item expanded "><a href="../cropseq/lazy.html"><strong aria-hidden="true">2.3.</strong> Background Processing</a></li><li class="chapter-item expanded "><a href="../cropseq/processing.html"><strong aria-hidden="true">2.4.</strong> Sequence Processing</a></li><li><ol class="section"><li class="chapter-item expanded "><a href="../cropseq/directory.html"><strong aria-hidden="true">2.4.1.</strong> Directory Management</a></li><li class="chapter-item expanded "><a href="../cropseq/metadata.html"><strong aria-hidden="true">2.4.2.</strong> Metadata</a></li><li class="chapter-item expanded "><a href="../cropseq/10x_alignment.html" class="active"><strong aria-hidden="true">2.4.3.</strong> 10X Alignment</a></li><li class="chapter-item expanded "><a href="../cropseq/pcr_alignment.html"><strong aria-hidden="true">2.4.4.</strong> PCR Alignment</a></li><li class="chapter-item expanded "><a href="../cropseq/assignment.html"><strong aria-hidden="true">2.4.5.</strong> Assignment</a></li><li class="spacer"></li></ol></li></ol></li></ol>
            </div>
            <div id="sidebar-resize-handle" class="sidebar-resize-handle"></div>
        </nav>

        <div id="page-wrapper" class="page-wrapper">

            <div class="page">
                <div id="menu-bar-hover-placeholder"></div>
                <div id="menu-bar" class="menu-bar sticky bordered">
                    <div class="left-buttons">
                        <button id="sidebar-toggle" class="icon-button" type="button" title="Toggle Table of Contents" aria-label="Toggle Table of Contents" aria-controls="sidebar">
                            <i class="fa fa-bars"></i>
                        </button>
                        <button id="theme-toggle" class="icon-button" type="button" title="Change theme" aria-label="Change theme" aria-haspopup="true" aria-expanded="false" aria-controls="theme-list">
                            <i class="fa fa-paint-brush"></i>
                        </button>
                        <ul id="theme-list" class="theme-popup" aria-label="Themes" role="menu">
                            <li role="none"><button role="menuitem" class="theme" id="light">Light (default)</button></li>
                            <li role="none"><button role="menuitem" class="theme" id="rust">Rust</button></li>
                            <li role="none"><button role="menuitem" class="theme" id="coal">Coal</button></li>
                            <li role="none"><button role="menuitem" class="theme" id="navy">Navy</button></li>
                            <li role="none"><button role="menuitem" class="theme" id="ayu">Ayu</button></li>
                        </ul>
                        <button id="search-toggle" class="icon-button" type="button" title="Search. (Shortkey: s)" aria-label="Toggle Searchbar" aria-expanded="false" aria-keyshortcuts="S" aria-controls="searchbar">
                            <i class="fa fa-search"></i>
                        </button>
                    </div>

                    <h1 class="menu-title">How Do I...?</h1>

                    <div class="right-buttons">
                        <a href="../print.html" title="Print this book" aria-label="Print this book">
                            <i id="print-button" class="fa fa-print"></i>
                        </a>
                    </div>
                </div>

                <div id="search-wrapper" class="hidden">
                    <form id="searchbar-outer" class="searchbar-outer">
                        <input type="search" id="searchbar" name="searchbar" placeholder="Search this book ..." aria-controls="searchresults-outer" aria-describedby="searchresults-header">
                    </form>
                    <div id="searchresults-outer" class="searchresults-outer hidden">
                        <div id="searchresults-header" class="searchresults-header"></div>
                        <ul id="searchresults">
                        </ul>
                    </div>
                </div>
                <!-- Apply ARIA attributes after the sidebar and the sidebar toggle button are added to the DOM -->
                <script type="text/javascript">
                    document.getElementById('sidebar-toggle').setAttribute('aria-expanded', sidebar === 'visible');
                    document.getElementById('sidebar').setAttribute('aria-hidden', sidebar !== 'visible');
                    Array.from(document.querySelectorAll('#sidebar a')).forEach(function(link) {
                        link.setAttribute('tabIndex', sidebar === 'visible' ? 0 : -1);
                    });
                </script>

                <div id="content" class="content">
                    <main>
                        <h1 id="10x-alignment"><a class="header" href="#10x-alignment">10X Alignment</a></h1>
<p>This section will cover the sequence alignment steps for the 10X sequences.
It assumes you have a similar directory structure to the one described in
the previous sections.</p>
<h2 id="installation"><a class="header" href="#installation">Installation</a></h2>
<p>We are going to use <code>kb-python</code> for sequencing alignment, which is a wrapper
of the <code>kallisto</code> and <code>bustools</code> commandline tools.</p>
<pre><code class="language-bash"># enter our conda environment (can skip if already in environment)
conda activate my_cropseq

# install kb-python via pip
pip install kb-python

# check if installation worked
kb --version
</code></pre>
<h2 id="creating-your-results-directory"><a class="header" href="#creating-your-results-directory">Creating your results directory</a></h2>
<p><code>kb</code> will create an output directory <strong>for each of the samples</strong>.
This quickly gets unwieldy, so I recommend creating a new directory
in your project to hold these data for easy access later.</p>
<p>From the <strong>root of your project directory</strong> run:</p>
<pre><code class="language-bash">mkdir -p alignment/10X
</code></pre>
<p>Your project directory should now look like the following:</p>
<pre><code class="language-txt">project_name/
└── meta/
    └── whitelist/
        └── 3M-february-2018.txt.gz
    └── 10X_index/
        ├── index.idx 
        ├── t2g.txt
        ├── spliced_t2c.txt
        └── unspliced_t2c.txt
    └── pcr_index/
└── sequence/
    └── 10X/
        ├── sampleX_R1.fastq.gz
        ├── sampleX_R2.fastq.gz
        :        ...
        ├── sampleY_R1.fastq.gz
        └── sampleY_R2.fastq.gz
    └── pcr/
└── alignment/
    └── 10X/
</code></pre>
<h2 id="running-kb-count"><a class="header" href="#running-kb-count">Running <code>kb count</code></a></h2>
<p>We can now run the alignment step which will create the <code>cell x gene</code>
count matrix.</p>
<p>We will do this using the <code>kb count</code> submodule of <code>kb</code>.</p>
<pre><code class="language-bash">kb count \
  --verbose \
  --h5ad \
  -t 4 \
  -x 10xv3 \
  -w meta/whitelist/3M-february-2018.txt.gz \
  -g meta/10X_index/t2g.txt \
  -i meta/10X_index/index.idx \
  --workflow lamanno \
  --filter bustools \
  -c1 meta/10X_index/spliced_t2c.txt \
  -c2 meta/10X_index/unspliced_t2c.txt \
  -o alignment/10X/SampleX \
  SampleX_R1.fastq.gz SampleX_R2.fastq.gz
</code></pre>
<h2 id="running-multiple-sequencing-datasets-as-one-sample"><a class="header" href="#running-multiple-sequencing-datasets-as-one-sample">Running multiple sequencing datasets as one sample</a></h2>
<p>Sometimes you have the same sample split across multiple sequencing files.
<code>kb</code> can handle this easily, and the only change to the above command is to
include the path of all files under the same sample name at the end.</p>
<pre><code class="language-bash">kb count \
  --verbose \
  --h5ad \
  -t 4 \
  -x 10xv3 \
  -w meta/whitelist/3M-february-2018.txt.gz \
  -g meta/10X_index/t2g.txt \
  -i meta/10X_index/index.idx \
  --workflow lamanno \
  --filter bustools \
  -c1 meta/10X_index/spliced_t2c.txt \
  -c2 meta/10X_index/unspliced_t2c.txt \
  -o alignment/10X/SampleX \
  SampleX_1_R1.fastq.gz SampleX_1_R2.fastq.gz \
  SampleX_2_R1.fastq.gz SampleX_2_R2.fastq.gz \
  SampleX_3_R1.fastq.gz SampleX_3_R2.fastq.gz
</code></pre>
<h2 id="running-multiple-samples-at-once"><a class="header" href="#running-multiple-samples-at-once">Running multiple samples at once</a></h2>
<p>Parallel processing of multiple samples at once is outside the scope of this
tutorial, but if you are interested in efficient ways of doing this I
recommend checking out:</p>
<ul>
<li><a href="https://www.nextflow.io/">nextflow</a></li>
<li><a href="https://snakemake.readthedocs.io/en/stable/">snakemake</a></li>
</ul>
<p>Otherwise, you can take the above command and put it into a for-loop fairly easily:</p>
<pre><code class="language-bash">#!/bin/bash

for sample_id in sequence/10X/*R1.fastq.gz;
do
    basename=$(basename -s &quot;.fastq.gz&quot; $sample_id);
    r1=$sample_id;
    r2=${r1/R1/R2};

    kb count \
      --verbose \
      --h5ad \
      -t 4 \
      -x 10xv3 \
      -w meta/whitelist/3M-february-2018.txt.gz \
      -g meta/10X_index/t2g.txt \
      -i meta/10X_index/index.idx \
      --workflow lamanno \
      --filter bustools \
      -c1 meta/10X_index/spliced_t2c.txt \
      -c2 meta/10X_index/unspliced_t2c.txt \
      -o alignment/10X/${basename} \
      $r1 $r2
    
done
</code></pre>
<h2 id="results"><a class="header" href="#results">Results</a></h2>
<p>Once <code>kb</code> has finished we can take a look at the data and see
if it is as expected.</p>
<p>If you explore the directory <code>alignment/10X/&lt;your_sample_name&gt;</code> you should see
the following:</p>
<pre><code class="language-txt">.
├── 10xv3_whitelist.txt
├── counts_filtered
│   ├── adata.h5ad
│   ├── spliced.barcodes.txt
│   ├── spliced.genes.txt
│   ├── spliced.mtx
│   ├── unspliced.barcodes.txt
│   ├── unspliced.genes.txt
│   └── unspliced.mtx
├── counts_unfiltered
│   ├── adata.h5ad
│   ├── spliced.barcodes.txt
│   ├── spliced.genes.txt
│   ├── spliced.mtx
│   ├── unspliced.barcodes.txt
│   ├── unspliced.genes.txt
│   └── unspliced.mtx
├── filter_barcodes.txt
├── inspect.json
├── inspect.spliced.json
├── inspect.unspliced.json
├── kb_info.json
├── matrix.ec
├── output.bus
├── output.filtered.bus
├── output.unfiltered.bus
├── run_info.json
├── spliced.filtered.bus
├── spliced.unfiltered.bus
├── transcripts.txt
├── unspliced.filtered.bus
└── unspliced.unfiltered.bus
</code></pre>
<p>The <code>cell x gene</code> matrix will be under <code>counts_filtered/adata.h5ad</code> or <code>counts_unfiltered/adata.h5ad</code>.</p>
<p>You can decide which of these you will use downstream, though I
recommend using the <code>counts_filtered</code> version.</p>

                    </main>

                    <nav class="nav-wrapper" aria-label="Page navigation">
                        <!-- Mobile navigation buttons -->
                            <a rel="prev" href="../cropseq/metadata.html" class="mobile-nav-chapters previous" title="Previous chapter" aria-label="Previous chapter" aria-keyshortcuts="Left">
                                <i class="fa fa-angle-left"></i>
                            </a>
                            <a rel="next" href="../cropseq/pcr_alignment.html" class="mobile-nav-chapters next" title="Next chapter" aria-label="Next chapter" aria-keyshortcuts="Right">
                                <i class="fa fa-angle-right"></i>
                            </a>
                        <div style="clear: both"></div>
                    </nav>
                </div>
            </div>

            <nav class="nav-wide-wrapper" aria-label="Page navigation">
                    <a rel="prev" href="../cropseq/metadata.html" class="nav-chapters previous" title="Previous chapter" aria-label="Previous chapter" aria-keyshortcuts="Left">
                        <i class="fa fa-angle-left"></i>
                    </a>
                    <a rel="next" href="../cropseq/pcr_alignment.html" class="nav-chapters next" title="Next chapter" aria-label="Next chapter" aria-keyshortcuts="Right">
                        <i class="fa fa-angle-right"></i>
                    </a>
            </nav>

        </div>

        <script type="text/javascript">
            window.playground_copyable = true;
        </script>
        <script src="../elasticlunr.min.js" type="text/javascript" charset="utf-8"></script>
        <script src="../mark.min.js" type="text/javascript" charset="utf-8"></script>
        <script src="../searcher.js" type="text/javascript" charset="utf-8"></script>
        <script src="../clipboard.min.js" type="text/javascript" charset="utf-8"></script>
        <script src="../highlight.js" type="text/javascript" charset="utf-8"></script>
        <script src="../book.js" type="text/javascript" charset="utf-8"></script>

        <!-- Custom JS scripts -->
    </body>
</html>
