<!DOCTYPE HTML>
<html lang="en" class="sidebar-visible no-js light">
    <head>
        <!-- Book generated using mdBook -->
        <meta charset="UTF-8">
        <title>Assignment - How Do I...?</title>
        <!-- Custom HTML head -->
        <meta content="text/html; charset=utf-8" http-equiv="Content-Type">
        <meta name="description" content="">
        <meta name="viewport" content="width=device-width, initial-scale=1">
        <meta name="theme-color" content="#ffffff" />

        <link rel="icon" href="../favicon.svg">
        <link rel="shortcut icon" href="../favicon.png">
        <link rel="stylesheet" href="../css/variables.css">
        <link rel="stylesheet" href="../css/general.css">
        <link rel="stylesheet" href="../css/chrome.css">
        <link rel="stylesheet" href="../css/print.css" media="print">
        <!-- Fonts -->
        <link rel="stylesheet" href="../FontAwesome/css/font-awesome.css">
        <link rel="stylesheet" href="../fonts/fonts.css">
        <!-- Highlight.js Stylesheets -->
        <link rel="stylesheet" href="../highlight.css">
        <link rel="stylesheet" href="../tomorrow-night.css">
        <link rel="stylesheet" href="../ayu-highlight.css">

        <!-- Custom theme stylesheets -->
    </head>
    <body>
        <!-- Provide site root to javascript -->
        <script type="text/javascript">
            var path_to_root = "../";
            var default_theme = window.matchMedia("(prefers-color-scheme: dark)").matches ? "navy" : "light";
        </script>

        <!-- Work around some values being stored in localStorage wrapped in quotes -->
        <script type="text/javascript">
            try {
                var theme = localStorage.getItem('mdbook-theme');
                var sidebar = localStorage.getItem('mdbook-sidebar');

                if (theme.startsWith('"') && theme.endsWith('"')) {
                    localStorage.setItem('mdbook-theme', theme.slice(1, theme.length - 1));
                }

                if (sidebar.startsWith('"') && sidebar.endsWith('"')) {
                    localStorage.setItem('mdbook-sidebar', sidebar.slice(1, sidebar.length - 1));
                }
            } catch (e) { }
        </script>

        <!-- Set the theme before any content is loaded, prevents flash -->
        <script type="text/javascript">
            var theme;
            try { theme = localStorage.getItem('mdbook-theme'); } catch(e) { }
            if (theme === null || theme === undefined) { theme = default_theme; }
            var html = document.querySelector('html');
            html.classList.remove('no-js')
            html.classList.remove('light')
            html.classList.add(theme);
            html.classList.add('js');
        </script>

        <!-- Hide / unhide sidebar before it is displayed -->
        <script type="text/javascript">
            var html = document.querySelector('html');
            var sidebar = 'hidden';
            if (document.body.clientWidth >= 1080) {
                try { sidebar = localStorage.getItem('mdbook-sidebar'); } catch(e) { }
                sidebar = sidebar || 'visible';
            }
            html.classList.remove('sidebar-visible');
            html.classList.add("sidebar-" + sidebar);
        </script>

        <nav id="sidebar" class="sidebar" aria-label="Table of contents">
            <div class="sidebar-scrollbox">
                <ol class="chapter"><li class="chapter-item expanded affix "><a href="../intro.html">Introduction</a></li><li class="spacer"></li><li class="chapter-item expanded "><a href="../crispr_screen/intro.html"><strong aria-hidden="true">1.</strong> CRISPR Screening</a></li><li><ol class="section"><li class="chapter-item expanded "><a href="../crispr_screen/abundance.html"><strong aria-hidden="true">1.1.</strong> Counting sgRNA abundance</a></li><li><ol class="section"><li class="chapter-item expanded "><a href="../crispr_screen/reference.html"><strong aria-hidden="true">1.1.1.</strong> Reference Library</a></li><li class="chapter-item expanded "><a href="../crispr_screen/counting.html"><strong aria-hidden="true">1.1.2.</strong> Guide Counting</a></li></ol></li><li class="chapter-item expanded "><a href="../crispr_screen/expression.html"><strong aria-hidden="true">1.2.</strong> Differential Expression</a></li><li class="chapter-item expanded "><a href="../crispr_screen/visualization.html"><strong aria-hidden="true">1.3.</strong> Visualization</a></li></ol></li><li class="chapter-item expanded "><a href="../cropseq/intro.html"><strong aria-hidden="true">2.</strong> CROP-Seq</a></li><li><ol class="section"><li class="chapter-item expanded "><a href="../cropseq/installation.html"><strong aria-hidden="true">2.1.</strong> Installation</a></li><li class="chapter-item expanded "><a href="../cropseq/environment.html"><strong aria-hidden="true">2.2.</strong> Environment</a></li><li class="chapter-item expanded "><a href="../cropseq/lazy.html"><strong aria-hidden="true">2.3.</strong> Background Processing</a></li><li class="chapter-item expanded "><a href="../cropseq/processing.html"><strong aria-hidden="true">2.4.</strong> Sequence Processing</a></li><li><ol class="section"><li class="chapter-item expanded "><a href="../cropseq/directory.html"><strong aria-hidden="true">2.4.1.</strong> Directory Management</a></li><li class="chapter-item expanded "><a href="../cropseq/metadata.html"><strong aria-hidden="true">2.4.2.</strong> Metadata</a></li><li class="chapter-item expanded "><a href="../cropseq/10x_alignment.html"><strong aria-hidden="true">2.4.3.</strong> 10X Alignment</a></li><li class="chapter-item expanded "><a href="../cropseq/pcr_alignment.html"><strong aria-hidden="true">2.4.4.</strong> PCR Alignment</a></li><li class="chapter-item expanded "><a href="../cropseq/assignment.html" class="active"><strong aria-hidden="true">2.4.5.</strong> Assignment</a></li><li class="spacer"></li></ol></li></ol></li></ol>
            </div>
            <div id="sidebar-resize-handle" class="sidebar-resize-handle"></div>
        </nav>

        <div id="page-wrapper" class="page-wrapper">

            <div class="page">
                <div id="menu-bar-hover-placeholder"></div>
                <div id="menu-bar" class="menu-bar sticky bordered">
                    <div class="left-buttons">
                        <button id="sidebar-toggle" class="icon-button" type="button" title="Toggle Table of Contents" aria-label="Toggle Table of Contents" aria-controls="sidebar">
                            <i class="fa fa-bars"></i>
                        </button>
                        <button id="theme-toggle" class="icon-button" type="button" title="Change theme" aria-label="Change theme" aria-haspopup="true" aria-expanded="false" aria-controls="theme-list">
                            <i class="fa fa-paint-brush"></i>
                        </button>
                        <ul id="theme-list" class="theme-popup" aria-label="Themes" role="menu">
                            <li role="none"><button role="menuitem" class="theme" id="light">Light (default)</button></li>
                            <li role="none"><button role="menuitem" class="theme" id="rust">Rust</button></li>
                            <li role="none"><button role="menuitem" class="theme" id="coal">Coal</button></li>
                            <li role="none"><button role="menuitem" class="theme" id="navy">Navy</button></li>
                            <li role="none"><button role="menuitem" class="theme" id="ayu">Ayu</button></li>
                        </ul>
                        <button id="search-toggle" class="icon-button" type="button" title="Search. (Shortkey: s)" aria-label="Toggle Searchbar" aria-expanded="false" aria-keyshortcuts="S" aria-controls="searchbar">
                            <i class="fa fa-search"></i>
                        </button>
                    </div>

                    <h1 class="menu-title">How Do I...?</h1>

                    <div class="right-buttons">
                        <a href="../print.html" title="Print this book" aria-label="Print this book">
                            <i id="print-button" class="fa fa-print"></i>
                        </a>
                    </div>
                </div>

                <div id="search-wrapper" class="hidden">
                    <form id="searchbar-outer" class="searchbar-outer">
                        <input type="search" id="searchbar" name="searchbar" placeholder="Search this book ..." aria-controls="searchresults-outer" aria-describedby="searchresults-header">
                    </form>
                    <div id="searchresults-outer" class="searchresults-outer hidden">
                        <div id="searchresults-header" class="searchresults-header"></div>
                        <ul id="searchresults">
                        </ul>
                    </div>
                </div>
                <!-- Apply ARIA attributes after the sidebar and the sidebar toggle button are added to the DOM -->
                <script type="text/javascript">
                    document.getElementById('sidebar-toggle').setAttribute('aria-expanded', sidebar === 'visible');
                    document.getElementById('sidebar').setAttribute('aria-hidden', sidebar !== 'visible');
                    Array.from(document.querySelectorAll('#sidebar a')).forEach(function(link) {
                        link.setAttribute('tabIndex', sidebar === 'visible' ? 0 : -1);
                    });
                </script>

                <div id="content" class="content">
                    <main>
                        <h1 id="assignment"><a class="header" href="#assignment">Assignment</a></h1>
<p>Now that we have the PCR sequences aligned to their respective guides
we can perform the assignment step.</p>
<h2 id="summary"><a class="header" href="#summary">Summary</a></h2>
<p>The assignment step will be done with <a href="https://github.com/noamteyssier/geomux"><code>geomux</code></a>.</p>
<p>This is a tool I created based on a hypergeometric test which can assign cells
to guides with high confidence.</p>
<p>The intuition behind the mathematics is that every cell with an sgRNA will
have a background distribution (or sequencing noise) and a signal distribution
(i.e. its true sgRNAs).
The hypergeometric test is used to identify what sgRNA abundance occurs above
random chance, and then a ratio test is used to select for cells with only a
single signal sgRNA (i.e. filter against high multiplicity of infection).</p>
<h2 id="installation"><a class="header" href="#installation">Installation</a></h2>
<p>We can install <code>geomux</code> directly from its github repo.</p>
<pre><code class="language-bash">pip install git+https://github.com/noamteyssier/geomux
</code></pre>
<h2 id="usage"><a class="header" href="#usage">Usage</a></h2>
<p>We can use <code>geomux</code> directly from the command line, or we can use it from
a jupyter notebook.</p>
<p>I prefer to run it from the commandline and will include those instructions
here.</p>
<p>If you prefer to use it from a jupyter notebook, check out the
<a href="https://github.com/noamteyssier/geomux#python-module">tutorial</a>
on the repo.</p>
<h3 id="creating-an-assignment-directory"><a class="header" href="#creating-an-assignment-directory">Creating an assignment directory</a></h3>
<p>First we create a directory to hold our assignment results.</p>
<pre><code class="language-bash">mkdir -p meta/assignment/
</code></pre>
<h3 id="running-geomux"><a class="header" href="#running-geomux">Running <code>geomux</code></a></h3>
<p>We can now run <code>geomux</code> with the following command:</p>
<pre><code class="language-bash">geomux \
    -i alignment/pcr/sampleX/counts_filtered/adata.h5ad \
    -o meta/assignment/sampleX.tab
</code></pre>
<p>This will create a tab-separated file (<code>*.tab</code>) which will have the cell-barcode,
associated guide, and statistics about the assignment provided.</p>
<h2 id="intersecting-the-10x-sequencing-data-and-the-assignments"><a class="header" href="#intersecting-the-10x-sequencing-data-and-the-assignments">Intersecting the 10X sequencing data and the Assignments</a></h2>
<p>We have now done all the steps necessary to intersect the 10X single-cell sequencing
data with the enrichment PCR.</p>
<p>We can now just load in the 10X sequencing data and merge the dataframes together
to have our cells assigned to their respective sgRNAs.</p>
<p>The next step will assume you have <a href="https://scanpy.readthedocs.io/"><code>scanpy</code></a>
installed already.</p>
<pre><code class="language-python">import pandas as pd
import scanpy as sc

# load assignments
assignments = pd.read_csv(
    &quot;meta/assignment/sampleX.tab&quot;, 
    sep=&quot;\t&quot;, 
    index_col=&quot;barcode&quot;
)

# load anndata (10X)
adata = sc.read(
    &quot;alignment/10x/sampleX/counts_filtered/adata.h5ad&quot;
)

# merge the two
adata.obs = adata.obs.merge(
    assignment, 
    left_index=True, 
    right_index=True, 
    how=&quot;left&quot;
)

# drop cells that do not have an assignment
adata = adata.obs[~adata.obs.guide.isna()].copy()

# view the merged anndata
adata
</code></pre>

                    </main>

                    <nav class="nav-wrapper" aria-label="Page navigation">
                        <!-- Mobile navigation buttons -->
                            <a rel="prev" href="../cropseq/pcr_alignment.html" class="mobile-nav-chapters previous" title="Previous chapter" aria-label="Previous chapter" aria-keyshortcuts="Left">
                                <i class="fa fa-angle-left"></i>
                            </a>
                        <div style="clear: both"></div>
                    </nav>
                </div>
            </div>

            <nav class="nav-wide-wrapper" aria-label="Page navigation">
                    <a rel="prev" href="../cropseq/pcr_alignment.html" class="nav-chapters previous" title="Previous chapter" aria-label="Previous chapter" aria-keyshortcuts="Left">
                        <i class="fa fa-angle-left"></i>
                    </a>
            </nav>

        </div>

        <script type="text/javascript">
            window.playground_copyable = true;
        </script>
        <script src="../elasticlunr.min.js" type="text/javascript" charset="utf-8"></script>
        <script src="../mark.min.js" type="text/javascript" charset="utf-8"></script>
        <script src="../searcher.js" type="text/javascript" charset="utf-8"></script>
        <script src="../clipboard.min.js" type="text/javascript" charset="utf-8"></script>
        <script src="../highlight.js" type="text/javascript" charset="utf-8"></script>
        <script src="../book.js" type="text/javascript" charset="utf-8"></script>

        <!-- Custom JS scripts -->
    </body>
</html>
